File: solutions/php/0019-remov-nth-node-from-end-of-list.php
<?php

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution
{
    /**
     * @param ListNode $head
     * @param int $n
     * @return ListNode
     */
    public function removeNthFromEnd(ListNode $head, int $n): ?ListNode
    {
        $newHead = new ListNode(next: $head);
        $nth = $newHead;
        $len = 0;

        while ($head = $head->next) {
            if (++$len < $n) {
                continue;
            }
            $nth = $nth->next;
        }
        $nth->next = $nth->next->next;

        return $newHead->next;
    }
}

File: solutions/go/0019-remov-nth-node-from-end-of-list.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	nth := &ListNode{
		Next: head,
	}
	newHead := nth

	for len, temp := 1, head.Next; temp != nil; len, temp = len+1, temp.Next {
		if len < n {
			continue
		}
		nth = nth.Next
	}
	nth.Next = nth.Next.Next

	return newHead.Next
}

// Another solution
func removeNthFromEnd2(head *ListNode, n int) *ListNode {
	nth := head
	len := 1

	for temp := head.Next; temp != nil; len, temp = len+1, temp.Next {
		if len <= n {
			continue
		}
		nth = nth.Next
	}

	if len == n {
		head = head.Next
	} else if nth.Next != nil {
		nth.Next = nth.Next.Next
	}

	return head
}
